AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStoreAdmin/test/SystemKey/SignAndVerify.dfy

```diff
@@ -0,0 +1,181 @@
+// - Sign an arbitrary content
+// - Verify an arbitrary content
+// - Fail to verify tampered content
+
+module {:options "/functionSyntax:4" } TestContentHandler {
+  import opened Wrappers
+  import Fixtures
+  import AdminFixtures
+  import KMS = Com.Amazonaws.Kms
+  import ContentHandler = SystemKey.ContentHandler
+  import Structure
+  import UTF8
+
+  const MutationCommitmentContentToSHA :=
+    map[
+      UTF8.EncodeAscii(Structure.HIERARCHY_VERSION) := UTF8.EncodeAscii(Structure.HIERARCHY_VERSION_VALUE),
+      UTF8.EncodeAscii(Structure.KEY_CREATE_TIME) := UTF8.EncodeAscii("now"),
+      UTF8.EncodeAscii(Structure.M_ORIGINAL) := UTF8.EncodeAscii("system-key-does-not-validate-original-content"),
+      UTF8.EncodeAscii(Structure.M_TERMINAL) := UTF8.EncodeAscii("system-key-does-not-validate-terminal-content"),
+      UTF8.EncodeAscii(Structure.M_INPUT) := UTF8.EncodeAscii("system-key-does-not-validate-input-content")
+    ]
+
+  const MutationCommitmentContent :=
+    ContentHandler.Content(
+      ContentToSHA := MutationCommitmentContentToSHA,
+      PartitionValue := "a-branch-key-id",
+      SortValue := Structure.MUTATION_COMMITMENT_TYPE,
+      UUIDValue := "a-uuid")
+
+  const TamperedMutationCommitmentContent :=
+    ContentHandler.Content(
+      ContentToSHA := MutationCommitmentContentToSHA,
+      PartitionValue := "a-branch-key-id",
+      SortValue := Structure.MUTATION_COMMITMENT_TYPE,
+      UUIDValue := "b-uuid")
+
+  const kmsId: string := Fixtures.publicKeyArn
+
+  lemma TestValuesAreValid()
+    ensures MutationCommitmentContent.ValidState()
+    ensures KMS.Types.IsValid_KeyIdType(kmsId)
+    ensures TamperedMutationCommitmentContent.ValidState()
+  {}
+
+  method ActualSign(
+    input: ContentHandler.SignInput,
+    logPrefix: string
+  )
+    returns (output: Result<KMS.Types.CiphertextType, ContentHandler.SignError>)
+    requires input.ValidState()
+    ensures input.ValidState()
+    modifies input.KmsTuple.Modifies
+    modifies input.Crypto.Modifies
+  {
+    assert input.ValidState();
+    var output? := ContentHandler.SignContent(input);
+    // These prints are helpful for debugging
+    // print logPrefix + " Attempted to Sign was succesful? " + if output?.Success? then "Yes" else "No" + " .\n";
+    // if (output?.Failure?) {
+    //   print logPrefix + " Error\t";
+    //   print output?.error;
+    //   print "\n";
+    // } else {
+    //   print logPrefix + " Result:\t";
+    //   print output?.value;
+    //   print "\n";
+    // }
+    return output?;
+  }
+
+  method ActualVerify(
+    input: ContentHandler.VerifyInput,
+    logPrefix: string
+  )
+    returns (output: Result<bool, ContentHandler.VerifyError>)
+    requires input.ValidState()
+    ensures input.ValidState()
+    modifies input.KmsTuple.Modifies
+    modifies input.Crypto.Modifies
+  {
+    assert input.ValidState();
+    // assume {:axiom} false; // Tony cannot figure out why the Modifies clause is violated
```

Comment:
since this is commented out, did tony figure it out?

Reply:
Yes I did! Thanks to this really bright wonderful guy named José.